### ISSUE TITLE
Arrow Flight RPC handling concurrent requests

### ISSUE THREAD

**Issue by pvardanis** [apache/arrow#40360]
### Describe the usage question you have. Please include as many useful details as  possible.

Continuing the [discussion](https://github.com/apache/arrow/issues/34607) around concurrent requests in `Arrow Flight RPC` on a new thread.

@lidavidm stated the following regarding concurrent requests in `Arrow Flight RPC` in Python.

> If you make 10 concurrent requests to a Python service, the service can only handle one at a time due to the GIL. So the concurrency will not get you any wall-clock-time speedup. @pvardanis notes that they got the same time whether they used a thread pool or ran the requests sequentially. That is the question I am answering. They've simply hijacked this thread to ask whether asyncio would make a difference: I contend it would not.

(cc @pitrou) I'm still confused about what happens when I send concurrent requests using a `ThreadPoolExecutor` with `max_workers=16` since my server seems to handle requests in parallel, in contrast to what you're stating:
```
05-Mar-24 11:58:29 - root - INFO: [📡] Starting server on `grpc://0.0.0.0:8080`...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - INFO: Processing data...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
05-Mar-24 11:58:43 - mac.service.arrow_flight.server - DEBUG: Starting batch processing...
```
The server is hosting a ML model of around 6GB in size and the prediction step is CPU-bound. It's still not clear to me what happens with the GIL and whether Arrow Flight RPC can process requests in parallel or not. 

What I've also noticed is that the higher the `max_workers` value and the number of concurrent requests I'm sending (e.g. 100) the more likely the server would hang for a bit during the prediction step, probably because CPU usage is at maximum. FWIW the results are not lost and I'm still getting what I'm supposed to, it just gets delayed a little bit more which ofc compared to handling the requests in parallel is way faster.

Here are some benchmarks I did:
<img width="302" alt="image" src="https://github.com/apache/arrow/assets/37624791/99b43e12-39ef-4759-a040-a84a7b2c5997">
`max_workers=1` shows the exact same behavior as sending the requests sequentially. In all benchmarks, parallel requests are significantly faster than sequential ones (i.e. with `max_workers=1`), but it seems there's a sweet spot of `max_workers` (which I assume corresponds to the number of concurrent requests I'm sending at once) where allocating too many workers pushes the CPU to its limits, thus the server hanging which results to slower elapsed time.

I have some other questions in addition:
- Can I configure the number of concurrent requests my server handles at a time via Arrow Flight RPC?
- Is it possible to split the server to multiple workers as if I'd have different instances of the server (including the model) running in parallel?

### Component(s)

FlightRPC, Python

**Comment by lidavidm** [comment]
Sorry, I was being a little unclear. I meant that for a CPU-bound service, effectively only one RPC can get serviced at a time (but they will run concurrently on separate threads) due to the GIL. However, if your CPU-bound operation is releasing the GIL, then that doesn't apply.

**Comment by lidavidm** [comment]
> Can I configure the number of concurrent requests my server handles at a time via Arrow Flight RPC?

No. gRPC doesn't expose meaningful control over this. Handle it at the application layer (e.g. a semaphore or a mutex+condition variable+counter).

> Is it possible to split the server to multiple workers as if I'd have different instances of the server (including the model) running in parallel?

Not really. If you host on say Kubernetes or otherwise have control over DNS, if you have your client connect to a host that resolves to multiple IP addresses, gRPC can distribute requests over those hosts. https://github.com/grpc/grpc/blob/master/doc/load-balancing.md#pick_first

(The wording is extremely confusing because it's overly abstracted and they also tried to invent their own load balancing API standard that everything is written in terms of, despite being deprecated and never really used)

Otherwise, use a proxy/load balancer in the middle.

**Comment by pvardanis** [comment]
@lidavidm Can I have a `ServerMiddleware` that uses an `asyncio.Semaphore` configured to a desired number for `max_concurrency`and run `do_exchange()` on the server side based on whether the semaphore is locked or not?

**Comment by lidavidm** [comment]
In this case, the semaphore would simply block the RPC entirely, the RPC will not continue until the middleware stops blocking.

**Comment by pitrou** [comment]
Right, and also, **asyncio primitives are not thread-safe, therefore they should not be used for OS thread synchronization**. Coming straight from the documentation here:
https://docs.python.org/3/library/asyncio-sync.html

**Comment by pvardanis** [comment]
> In this case, the semaphore would simply block the RPC entirely, the RPC will not continue until the middleware stops blocking.

If my intention is to have a limit on the concurrent requests I'm processing at once, I guess that's fine, no? In any case, if you think there's a more appropriate way to do something similar I'm open to suggestions.

**Comment by lidavidm** [comment]
It should be fine, I can't say I've tried this before but I don't think you have many other alternatives. You may want to consider something more sophisticated if you expect load to be wildly variable, e.g. rejecting requests with UNAVAILABLE and having clients retry later (or on a different host), since the request will still sit there and take up a server thread.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 365 days. Please remove the stale label or comment below, or this issue will be closed in 14 days. If this usage question has evolved into a feature request or docs update, please remove the 'Type: usage' label and add the 'Type: enhancement' label instead.
